README.md
# Shop [![Build Status](https://secure.travis-ci.org/romainberger/shop.png)](http://travis-ci.org/romainberger/shop) [![Gem Version](https://badge.fury.io/rb/shop.png)](http://badge.fury.io/rb/shop)

CLI for easier [PrestaShop](http://www.prestashop.com/en/) development. Actually it's just for the lazy people.

I always use Rakefiles and Makefiles for lots of tasks. With this gem you can use these tasks without having to create any *akefiles. For more infos about how and why see: [Introducing Shop](http://romainberger.com/2013/09/29/introducing-shop/).

Shop has been developed to work with PrestaShop 1.5 and 1.4 (support for 1.6 will be added once it's realeased).

## Installation

    gem install shop

## Usage

### New project

Creates a new Prestashop project with:

    shop new <directory> [version]

    # example:
    shop new
    # Download the framework (stable version) in the current directory

    shop new new-store
    # Download the framework in a new directory `new-store`

Download the latest stable version (1.5) of PrestaShop in the current directory or in a subdirectory if you provide the name as argument.

Even if the 1.6 version is not out yet you can create new projects with this version by passing the version as an argument:

    shop new 1.6
    # Create a new project in the current directory with Prestashop 1.6

    shop new new-store 1.6
    # Create a new project in a new directory `new-store` with the version 1.6

**Note**: Shop can work with an existing PrestaShop project, all you need is to initialize the Shop project (see [Initialization](https://github.com/romainberger/shop/#initialization) below)

### Installation

Since 1.5.4 [PrestaShop provides a CLI](http://doc.prestashop.com/display/PS15/Installing+PrestaShop+using+the+command+line) to install your PrestaShop. You can try to remember every arguments, or just use the Shop task and stay relatively sane (no offense to the PrestaShop developers but... come on). You will need to create the database before running this task.

    # php install/cli.php --name=Romain --lol=whyamitypingallthis
    # nope
    shop install

**Note**: the install task actually runs the PrestaShop PHP CLI. It just makes you happier with a nice prompt.

### Initialization

For some tasks, Shop will need to know the name of the theme you're using. To do that you need to init the project with:

    shop init <theme-name>

    # example:
    shop init my-theme

### Overrides

Creates overrides files for controllers and classes.

    shop override <controller|class> <name> [admin]

    # examples:

    # generate an override for the Product controller
    shop override controller Product

    # by default the controller will be placed in override/controllers/admin. To create an override in
    # override/controllers/admin add the `admin` argument:
    shop override controller AdminProducts admin

    # generate an override for the Product class
    shop override class Product

### New Pages

You can generate new pages for PrestaShop easily with the `controller` command. This will create the controller, template, CSS and JS files you need to create a new page.

    shop controller <page-name>

    # example:
    shop controller CustomPage

### Modules

**The modules related tasks will need the Shop project to be [initialized](https://github.com/romainberger/shop)**

Creates a new module (with simple module boilerplate):

    shop module <name>

    # example:
    shop module dinozaure

Creates a module template:

    shop module template <module-name> <hook-name>

    # example:
    shop module template blockcategories blockcategories

Creates a module css file in your theme:

    shop module css <module-name>

    # example:
    shop module css blockcategories

### Cache

Cleans the css and js caches.

    shop clean [cache]
    # cleans the css and js caches

When you create a new override for a controller or a class, the class index needs to be rewritten. You can call this task

    shop clean class
    # clean the class index - automatically done when generating a new controller or class

By default the `clean` task will clean the cache so you can just use `shop clean`.

**Note**: The class index is automatically regenerated when you create an override with Shop. You probably won't need to run this command.

### Jshint

Run jshint on the theme's files.

    shop jshint

To run jshint on the theme's files and the modules' files, run it with the `modules` argument:

    shop jshint modules

To run jshint on **every** files run it with the `hard` argument (prepare yourself for a crapload of errors):

    shop jshint hard

### Makefile

Okay I kinda lied. I still use a Makefile, but only for production servers, as they rarely run gems and I don't necessarily want to install a tool for a limited use (Shop is primarily built for development purpose). But some of these tasks can be usefull on production servers. Shop can generate a Makefile file that will allow you to run some tasks.

**The Makefile task will need the Shop project to be [initialized](https://github.com/romainberger/shop)**

If you already use a Makefile, the tasks will be added to it. If you don't, it will create a Makefile.

    shop makefile

You can then use these tasks. Their effect are similar to the Shop tasks.

    make clean-cache

    make clean-class


### Help

You can get a list of the available commands with the help:

    shop help

## Configuration

You can use a configuration file to automate some values in the templates files created (modules and for the installation of PrestaShop).
You can edit this file with the command

    shop edit

It will open the config file in your `$EDITOR`. It's YAML so it shouldn't be scary.

## Custom templates

Some tasks generates files, using templates located in [/templates](https://github.com/romainberger/shop/tree/master/templates). You can override these files to use custom ones. See the [template readme for more infos](https://github.com/romainberger/shop/tree/master/templates).

## Autocompletion

To enable autocompletion with Shop, [see the docs.](https://github.com/romainberger/shop/tree/master/completion)

## Compatibility

Shop has been developed and tested on osx. It may not work on platforms other than *nix, particularly the `new` command that uses the unix `unzip` command. I plan on working on rewriting this part with only Ruby so it could work everywhere. I'm not going to try it on Windows so feel free to open issues (or even better, contribute) if you notice something not working as expected.

## Who

By [Romain Berger](http://romainberger.com).
Ping me at [@romain__berger](http://twitter.com/romain__berger) if you're having issues or create an issue.

If you become billionaire with a shop made with the help of Shop you must buy me a beer.


Changelog.md
# Shop changelog

## Head
- remove the mysql2 dependency to avoid bugs
- improve jshint task to run on mobile theme
- change `clean class-index` task to only remove the file - re-creating the file sometimes prevent the server to write it
- change `new` task to be able to choose between Prestashop 1.5 or 1.6

## 0.1.9
- Change template system to use `<%= %>` instead of `{{}}`
- Creates the database in the `install` task
- Fuck yeah autocompletion

## 0.1.8
- By default the clean task clean the cache
- Add a `template` section to the configuration file so you can use custom templates
- Output the files created for every tasks
- Allow the use of the `module` task with or without file extension. ie. `shop module template name template` or `shop module template name template.tpl`
- Fix controller task and keep camelcased controllers name

## 0.1.6
- Add an `edit` command to edit the config file
- New task `controller` to create new pages
- Add some colors

## 0.1.5
- Configuration file

## 0.1.4
- First version


lib/shop/command.rb
# coding: utf-8

module Shop
  class Command
    class << self
      include Shop::Color

      def template
        Template.new
      end

      # Returns the theme name
      def theme
        if init?
          return File.read('.shop')
        else
          puts "#{red("Error")}: Project not initialized."
          puts "Please run #{cyan("shop init <theme-name>")}"
          exit
        end
      end

      # Returns the version of Prestashop installed
      def psVersion
        if !installed?
          puts "PrestaShop does not seem to be installed"
          exit
        end

        line = File.read("config/settings.inc.php")[/^\s*define\('_PS_VERSION_',\s*.*\)/]
        line.match(/.*define\('_PS_VERSION_',\s*\s*['"](.*)['"]\)/)[1]
      end

      def execute(*args)
        command = args.shift
        major   = args.shift
        minor   = args.shift
        extra   = args.shift

        return help unless command
        dispatch(command, major, minor, extra)
      end

      def dispatch(command, major, minor, extra)
        return newProject(major, minor)        if command == 'new'
        return init(major)                     if command == 'init'
        return install                         if command == 'install'
        return shopModule(major, minor, extra) if command == 'module'
        return override(major, minor, extra)   if command == 'override'
        return controller(major)               if command == 'controller'
        return clean(major)                    if command == 'clean'
        return jshint(major)                   if command == 'jshint'
        return makefile                        if command == 'makefile'
        return edit                            if command == 'edit'
        return version                         if command == "-v"
        return version                         if command == "--version"
        return help                            if command == 'help'
        return complete(major)                 if command == 'cmplt'

        puts "\n#{red("Error")}: Command not found"
        return help
      end

      # Download the framework in the current dir
      # or a creates a dir if an argument is given
      def newProject(major, minor)
        # Make sure these cases work:
        # shop new directory 1.6
        # shop new directory
        # shop new 1.6
        begin
          version = Float(major)
          version = String(version)
        rescue
          path = major
        end

        if version.nil?
          begin
            version = Float(minor)
            version = String(version)
          rescue
            version = "1.5"
          end
        end

        unless path.nil?
          FileUtils.mkpath(path)
          puts "Creating new PrestaShop project in ./#{path}"
        else
          path = "./"
          puts "Creating new PrestaShop project"
        end

        version ||= "1.5"

        if !["1.5", "1.6"].include?(version)
          print "Wrong version #{version}"
          exit
        end

        puts "Downloading the framework... "
        print "Prestashop Version #{version}"
        if version == "1.6"
          url = "https://github.com/PrestaShop/PrestaShop/archive/1.6.zip"
        else
          url = "https://github.com/PrestaShop/PrestaShop-1.5/archive/master.zip"
        end

        open("master.zip", "wb") do |f|
          f << open(url).read
        end
        done

        # @todo unzip with a ruby way to avoid platform incompatibilities
        print "Unzip... "
        system "unzip -q master.zip"
        done

        print "Copying... "
        FileUtils.cp_r(Dir["PrestaShop-master/*"], path)
        done

        print "Cleaning... "
        # remove useless files
        File.delete("master.zip")
        FileUtils.rm_rf("PrestaShop-master")

        done
      end

      # Returns if the framework is already installed
      def installed?
        File.exists?('config/settings.inc.php')
      end

      # Runs the Prestashop install CLI but with a nice prompt
      #
      # See http://doc.prestashop.com/display/PS15/Installing+PrestaShop+using+the+command+line
      def install
        if installed?
          puts "PrestaShop appears to be already installed"
          exit
        end

        puts "Please answer the following: "

        # Default values
        config = ShopConfig.new
        db_server = config.get('install', 'db_server', 'localhost')
        db_user = config.get('install', 'db_user', 'root')
        db_password = config.get('install', 'db_password')
        country = config.get('install', 'country', 'fr')
        firstname = config.get('install', 'firstname', false)
        lastname = config.get('install', 'lastname', false)
        password = config.get('install', 'password', '0123456789')
        email = config.get('install', 'email', false)

        entry = Hash.new

        entry[:domain]      = ask('Domain: ')
        entry[:db_name]     = ask('Database name: ')
        entry[:db_server]   = ask('Database server: ') { |q| q.default = db_server }
        entry[:db_user]     = ask('Database user: ') { |q| q.default = db_user }
        entry[:db_password] = ask('Database password: ') { |q| q.default = db_password }
        entry[:country]     = ask('Country: ') { |q| q.default = country }
        entry[:firstname]   = ask('Firstname: ') { |q| q.default = firstname if firstname }
        entry[:lastname]    = ask('Lastname: ') { |q| q.default = lastname if lastname }
        entry[:password]    = ask('Password: ') { |q| q.default = password }
        entry[:email]       = ask('Email: ') { |q| q.default = email if email }

        command = "php install-dev/index_cli.php "
        command << "--domain=#{entry[:domain]} "
        command << "--db_name=#{entry[:db_name]} "
        command << "--db_server=#{entry[:db_server]} "
        command << "--db_user=#{entry[:db_user]} "
        command << "--db_password=#{entry[:db_password]} "
        command << "--country=#{entry[:country]} "
        command << "--firstname=#{entry[:firstname]} "
        command << "--lastname=#{entry[:lastname]} "
        command << "--password=#{entry[:password]} "
        command << "--email=#{entry[:email]} "
        command << "--newsletter=0"

        # run the php script
        puts "Installing Prestashop please wait..."
        system command
        done
      end

      # Init the project
      def init(name)
        if name.nil?
          puts "#{red("Error:")} Please specify the name of the theme"
          return puts "    $ shop init <theme-name>"
        end

        puts "      #{green("create")} .shop"
        File.open('.shop', 'w') do |f|
          f.write(name)
        end

        done
      end

      # Check if the Shop project has been initialized
      #
      # Returns a boolean
      def init?
        File.exists?('.shop')
      end

      # Creates a module or a module template
      # prefixed with shop for obvious reasons
      #
      # The project needs to be initialized
      def shopModule(major, minor, extra)
        if major == 'template'
          theme
          path = "themes/#{theme}/modules"
          FileUtils.mkpath(path) unless File.directory?(path)

          path = "#{path}/#{minor}"
          FileUtils.mkpath(path) unless File.directory?(path)

          filepath = "#{path}/#{extra}"
          filepath = "#{filepath}.tpl" unless filepath.match(/\.tpl$/)
          if File.exists?(filepath)
            puts "#{red("Error:")} File already exists"
            exit
          else
            puts "      #{green("create")} #{filepath}"
            File.open(filepath, "w")
          end
        elsif major == 'css'
          theme
          # css
          path = "themes/#{theme}/css/modules/#{minor}"
          FileUtils.mkpath(path) unless File.directory?(path)

          filepath = "#{path}/#{minor}"
          filepath = "#{filepath}.css" unless filepath.match(/\.css$/)

          if File.exists?(filepath)
            puts "#{red("Error:")} File already exists"
            exit
          elsif
            puts "      #{green("create")} #{filepath}"
            File.open(filepath, "w")
          end
        else
          # create a module
          if File.directory?("modules/#{major}")
            return puts "Module #{major} already exists"
          else
            puts "      #{green("create")} modules/#{major}"
            FileUtils.mkpath("modules/#{major}")
            config = ShopConfig.new
            values = {
              "name_capitalize" => major.capitalize,
              "name" => major,
              "author" => config.get('module', 'author'),
              "tab" => config.get('module', 'tab')
            }
            content = template.template("module.php", values)
            puts "      #{green("create")} modules/#{major}/#{major}.php"
            File.open("modules/#{major}/#{major}.php", 'w') do |f|
              f.write(content)
            end

            # copy the logos
            logo = config.get('module', 'logo')
            if logo.length != 0
              if File.exists?("#{logo}.png")
                puts "      #{green("create")} modules/#{major}/logo.png"
                FileUtils.cp("#{logo}.png", "modules/#{major}/logo.png")
              end
              if File.exists?("#{logo}.gif")
                puts "      #{green("create")} modules/#{major}/logo.gif"
                FileUtils.cp("#{logo}.gif", "modules/#{major}/logo.gif")
              end
            end
          end
        end

        done
      end

      # Creates an override for controllers and classes
      def override(major, minor, extra)
        name = minor.capitalize

        if major == 'controller'
          # controller
          if !extra.nil? && extra == 'admin'
            side = 'admin'
          else
            side = 'front'
          end

          name = "#{name}Controller"
          filename = "#{name}.php"
          path = "override/controllers/#{side}/#{filename}"

        elsif major == 'class'
          # class
          filename = "#{name}.php"
          path = "override/classes/#{filename}"
        end

        content = "<?php\n\nclass #{name} extends #{name}Core {\n\n}\n"

        if !File.directory?('override')
          return puts "#{red("Error:")} You need to be at the root of your Prestashop site"
        end

        puts "      #{green("create")} #{path}"
        File.open(path, 'w') do |f|
          f.write(content)
        end

        clean("class")

        done
      end

      # Creates a new page with controller and files associated
      #
      # @todo adapt this to work with PS 1.4 (needs a .php file in the root dir)
      def controller(major)
        if major.nil?
          puts "#{red("Error")}: Please provide a name for the page"
          exit
        end
        theme
        version = psVersion
        version = version[0, 3]

        # convert the name to camelcase to hyphen
        name = major.split(/(?=[A-Z])/)
        name = name.join('-').downcase

        controller_name = "#{major}Controller"
        filename = "#{controller_name}.php"
        if version.to_f >= 1.5
          controller_path = "controllers/front/#{filename}"
        elsif version == "1.4"
          controller_path = "controllers/#{filename}"
        end

        if File.exists?(controller_path)
          puts "#{red("Error")}: Controller already exists"
          exit
        end

        # datas for templates
        datas = {
          'controller_name' => controller_name,
          'name' => name
        }

        # For PS 1.4 creates a php file a the root
        # and use the appropriate template
        if version[0, 3] == "1.4"
            content = template.template("page.php", datas)
            puts "      #{green("create")} #{name}.php"
            File.open("#{name}.php", "w") do |f|
              f.write(content)
            end
            controllerContent = template.template("controller-1.4.php", datas)
        else
          controllerContent = template.template("controller.php", datas)
        end

        puts "      #{green("create")} #{controller_path}"
        File.open(controller_path, "w") do |f|
          f.write(controllerContent)
        end

        files = [
          "themes/#{theme}/#{name}.tpl",
          "themes/#{theme}/css/#{name}.css",
          "themes/#{theme}/js/#{name}.js"
        ]

        files.each do |f|
          puts "      #{green("create")} #{f}"
          File.open(f, "w")
        end

        clean("class")

        done
      end

      # Clean cache or class index
      def clean(major)
        if major.nil? || major == 'cache'
          theme
          files = Dir["themes/#{theme}/cache/*.css"]
          files = files + Dir["themes/#{theme}/cache/*.js"]

          files.each do |f|
            File.delete(f)
          end
        elsif major == 'class'
          version = psVersion
          if version[0, 3] == "1.4"
            return
          end

          print "      Cleaning class index... "
          index = "cache/class_index.php"
          if File.exists?(index)
            File.delete(index)
          end
        end

        done
      end

      # Run jshint
      def jshint(major)
        theme
        directory = Dir["themes/#{theme}"]

        if major == 'modules' || major == 'hard'
          modules = Dir["modules"]
          directory = directory + modules
        end
        if major == 'hard'
          prestashop = Dir["js"]
          directory = directory + prestashop
        end

        directory.each do |f|
          system "jshint #{f}"
        end
      end

      # Create a Makefile or add some tasks to an existing one
      def makefile
        theme
        datas = { 'theme' => theme}
        content = template.template('Makefile', datas)
        if File.exists?("Makefile")
          File.open("Makefile", "a") do |f|
            f.write(content)
          end
        else
          File.open("Makefile", "w") do |f|
            puts "      #{green("create")} Makefile"
            f.write(content)
          end
        end

        done
      end

      def done
        puts "\n      #{green("✔ Done")}"
      end

      def edit
        config = ShopConfig.new
        puts config.edit
      end

      # Returns the version of Shop
      def version
        puts "Shop #{Shop::VERSION}"
      end

      # Prints all the command available
      def help
        text = %{
          Shop v.#{Shop::VERSION} -------------------------------

            shop new <directory>                   Download the framework
            shop install                           Install Prestashop

            shop init                              Creates a Shop config file

            shop module <name>                     Creates a new module
            shop module template <name> <hook>     Creates a template for a given module / hook
            shop module css <name>                 Creates an override for a module css

            shop controller <controller-name>      Creates a new page with a controller and
                                                   assets needed

            shop override controller <name>        Creates an override for a controller
            shop override controller <name> admin  Creates an override for an admin controller
            shop override class <name>             Creates an override for a class

            shop clean [cache]                     Cleans the css and js caches
            shop clean class                       Cleans the class index

            shop jshint                            Run jshint on the theme files
            shop jshint modules                    Run jshint on the theme and modules files
            shop jshint hard                       Run jshint on every files

            shop makefile                          Creates a Makefile or add tasks to an existing one

            shop edit                              Open the configuration file in your $EDITOR
            shop help                              Output this help text
            shop version                           Output the version of Shop you are using

          See the complete documentation at:
          https://github.com/romainberger/shop

        }.gsub(/^ {10}/, '')
        puts text
      end

      # Method for the autocompletion
      def complete(word)
        complete = %{
          new
          init
          install
          module
          module template
          override
          override controller
          override class
          controller
          clean
          clean class
          jshint
          makefile
        }
        puts complete
      end
    end
  end
end
